pkg/storage/bulk/sst_batcher_test.go

```diff
@@ -203,8 +206,97 @@ func runTestImport(t *testing.T, batchSize int64) {
 						t.Logf("got      %d\t%v\t%v", i, got[i].Key, got[i].Value)
 					}
 				}
-				t.Fatalf("got %+v expected %+v", got, expected)
+				t.Fatalf("got      %+v\nexpected %+v", got, expected)
 			}
 		})
 	}
 }
+
+type mockSender func(span roachpb.Span) error
+
+func (m mockSender) AddSSTable(ctx context.Context, begin, end interface{}, data []byte) error {
+	return m(roachpb.Span{Key: begin.(roachpb.Key), EndKey: end.(roachpb.Key)})
+}
+
+// TestAddBigSpanningSSTWithSplits tests a situation where a large
+// spanning SST is being ingested over a span with a lot of splits.
+func TestAddBigSpanningSSTWithSplits(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	if testing.Short() {
+		t.Skip("this test needs to do a larger SST to see the quadratic mem usage on retries kick in.")
+	}
+
+	const numKeys, valueSize, splitEvery = 500, 5000, 1
+
+	prefix := encoding.EncodeUvarintAscending(keys.MakeTablePrefix(uint32(100)), uint64(1))
+	key := func(i int) []byte {
+		return encoding.EncodeVarintAscending(append([]byte{}, prefix...), int64(i))
+	}
+
+	var splits []roachpb.Key
+
+	// Create a large SST.
+	w, err := engine.MakeRocksDBSstFileWriter()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer w.Close()
+	r, _ := randutil.NewPseudoRand()
+	buf := make([]byte, valueSize)
+	for i := 0; i < numKeys; i++ {
+		randutil.ReadTestdataBytes(r, buf)
+		if i%splitEvery == 0 {
+			splits = append(splits, key(i))
+		}
+		if err := w.Add(engine.MVCCKeyValue{
+			Key:   engine.MVCCKey{Key: key(i)},
+			Value: roachpb.MakeValueFromString(string(buf)).RawBytes,
+		}); err != nil {
+			t.Fatal(err)
+		}
+	}
+	sst, err := w.Finish()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Keep track of the memory.
+	getMem := func() uint64 {
+		var stats runtime.MemStats
+		runtime.ReadMemStats(&stats)
+		return stats.HeapInuse
+	}
+	var baseline, early, late uint64
+	baseline = getMem()
+	var totalAdditionAttempts int
+	mock := mockSender(func(span roachpb.Span) error {
+		totalAdditionAttempts++
+		for i := range splits {
+			if span.ContainsKey(splits[i]) && !span.Key.Equal(splits[i]) {
+				earlySplit := numKeys / 100
+				if i == earlySplit {
+					early = getMem() - baseline
```

Comment:
I think we want to get rid of `baseline`  and these `- baseline`s, since they assume that men usage only increases which might not be true -- if for any reason mem usage _dropped_ between establishing baseline once of these calls, this subtraction just overflows.